mitodl/mit-open
# Publish topic channels based on resource count

### What are the relevant tickets?
Closes https://github.com/mitodl/hq/issues/5082

### Description (What does it do?)
This PR
- Adds a `published` field on the `Channel` model.
- Topic channels are set to `published: False` when there there are no published learning resources with that topic. This happens via plugins that are triggered, e.g., by ETL functions.
- Updates Channel APIs to only return published channels
    - if a topic has no associated resources, it is still returned by `api/v1/topics/`, but its channel_url is null. 

### How can this be tested?
1. On `main`: Note that some topics have no associated resources, but they still show up in the UI and their channel pages are viewable. For example, the subtopic "Family Enterprise" of "Business & Management" has no associated resources.
    - Check on `main` that http://localhost:8062/topics lists empty topics 
    - Click the topic link to view the empty channel pages 
    - To find all empty topics, `./manage.py pgshell` and
        ```sql
        SELECT 
        	t.name, 
        	COUNT(*) FILTER (WHERE r.published) AS resource_count,
        	c.published, array_agg(DISTINCT r.etl_source) FILTER (WHERE r.published) AS etl_sources
        FROM learning_resources_learningresourcetopic t
        JOIN channels_channeltopicdetail d ON d.topic_id = t.id
        JOIN channels_channel c ON c.id = d.channel_id
        LEFT JOIN learning_resources_learningresource_topics rt ON t.id = rt.learningresourcetopic_id
        LEFT JOIN learning_resources_learningresource r ON rt.learningresource_id = r.id 
        GROUP BY c.id, t.name
        ORDER BY COUNT(*) FILTER(WHERE r.published);
        ```
2. Check out this branch and restart docker with `RUN_DATA_MIGRATIONS=True` in your env.
3. All "empty topic" channels should now be unpublished, and no longer show up in the UI.
4. To verify that topic channels are added/removed appropriately during ETL, you could, for example:
    1. Run `./manage.py backpopulate_micromasters_data --delete`.
    2. Check that "Business Analytics" topic channel has been unpublished. *This topic has only 1 resource and it is from Micromasters, so it should now be unpublished.*
    3. Run `./manage.py backpopulate_micromasters_data` 
    4. Check that "Business Analytics" has been republished
